src/plugins/stake-plugins/generic/policyAdapters/TarotPoolAdaptor.ts

```diff
@@ -0,0 +1,491 @@
+      // get liquidity quote
+      const amountToken0Min = token0Amount.mul(100).div(100 * SLIPPAGE_FACTOR)
+      const amountToken1Min = token1Amount.mul(100).div(100 * SLIPPAGE_FACTOR)
+
+      const [amountA, amountB, estimateMinLiquidity] = await velodromeRouterContract.quoteAddLiquidity(
+        token0.contractAddress,
+        token1.contractAddress,
+        isStable,
+        velodromeFactoryContractAddress,
+        token0Amount,
+        token0Amount
+      )
+
+      // approve lp token spending to router contract
+      const lpTokenAllowance = await lpTokenContract.allowance(walletAddress, tarotRouterContractAddress)
+      if (lpTokenAllowance.lt(estimateMinLiquidity)) {
+        txs.push(
+          async () =>
+            await lpTokenContract.populateTransaction.approve(tarotRouterContractAddress, estimateMinLiquidity, {
+              gasPrice,
+              gasLimit: '60000',
+              customData: {
+                name: metadataName,
+                category: 'Expense:Fees',
+                notes: `Approve ${lpToken.symbol} for ${metadataPoolAssetName} liquidity pool contract`
+              }
+            })
+        )
+      }
+
+      // create addLiquidity transaction
+      const deadline = Math.round(Date.now() / 1000) + DEADLINE_OFFSET
+      txs.push(
+        async () =>
+          await velodromeRouterContract.populateTransaction.addLiquidity(
+            token0.contractAddress,
+            token1.contractAddress,
+            isStable,
+            amountA,
+            amountB,
+            amountA.mul(100).div(100 * SLIPPAGE_FACTOR),
+            amountB.mul(100).div(100 * SLIPPAGE_FACTOR),
+            walletAddress,
+            deadline,
+            {
+              gasPrice,
+              gasLimit: '500000',
+              customData: {
+                name: metadataName,
+                category: 'Transfer:Staking',
+                notes: `Add liquidity to ${metadataPoolAssetName} liquidity pool contract`
+              }
+            }
+          )
+      )
+
+      // mint collateral
+      const tarotRouterContract = TarotRouter__factory.connect(tarotRouterContractAddress, provider)
+      txs.push(async (prevTx?: ethers.providers.TransactionResponse) => {
+        let addedLiquidityAmount = estimateMinLiquidity
+        if (prevTx != null) {
+          const receipt = await prevTx.wait(1)
+
+          const eventLogs = receipt.logs.filter(log => log.address.toLowerCase() === lpTokenContract.address.toLowerCase())
+          const log = eventLogs.filter(
+            log => log.topics[0] === lpTokenContract.interface.getEventTopic(lpTokenContract.interface.events['Transfer(address,address,uint256)'])
+          )[0]
+
+          if (log == null) throw new Error('Cannot find log with amount')
+
+          const decodedData = ethers.utils.defaultAbiCoder.decode(['uint256'], log.data)[0]
+
+          if (BigNumber.isBigNumber(decodedData)) {
+            addedLiquidityAmount = decodedData
```

Comment:
We don't know the amount because we can't call addLiquidity before approving the token. It's a safe overestimate to use because you'll always get less lp token than you approve due to slippage

Reply:
As long so you know that the estimate is always greater than the actual liquidity amount